src/appleseed/renderer/modeling/aov/cryptomatte.cpp

```diff
@@ -0,0 +1,908 @@
+                {
+                    const float weight = m_renderer_filter->evaluate(rx - dx, ry - dy);
+                    WeightMap& weight_map = m_pixel_samples[(m_tile_origin_y + ry) * m_frame_width + (m_tile_origin_x + rx)];
+                    const float old_value = weight_map.get(m3hash);
+                    weight_map.insert(m3hash, old_value + weight); // possible race conditions writing to pixels on the tile edges
+                }
+            }
+            return;
+        }
+    };
+
+    const char* CryptomatteObjectAOVModel = "cryptomatte_object_aov";
+    const char* CryptomatteMaterialAOVModel = "cryptomatte_material_aov";
+}
+
+
+//
+// CryptomatteAOV class implementation.
+//
+
+struct CryptomatteAOV::Impl
+{
+    vector<WeightMap>                   m_pixel_samples;
+    map<pair<int, int>, NameMap>        m_tile_name_map;
+    unique_ptr<Image>                   m_image;
+    size_t                              m_num_layers;
+    CryptomatteAOV::CryptomatteType     m_layer_type;
+};
+
+CryptomatteAOV::CryptomatteAOV(const ParamArray& params)
+  : AOV("cryptomatte", params)
+  , impl(new Impl())
+{
+    const string cryptomatte_type = params.get_optional<string>("cryptomatte_type", "object_names", make_vector("object_names" , "material_names"));
+    
+    if (cryptomatte_type == "object_names")
+    {
+        impl->m_layer_type = CryptomatteType::ObjectNames;
+        set_name(CryptomatteObjectAOVModel);
+    }
+    else if (cryptomatte_type == "material_names")
+    {
+        impl->m_layer_type = CryptomatteType::MaterialNames;
+        set_name(CryptomatteMaterialAOVModel);
+    }
+
+    impl->m_num_layers = params.get_optional<size_t>("cryptomatte_num_layers", 6);
+}
+
+CryptomatteAOV::~CryptomatteAOV()
+{
+    delete impl;
+}
+
+const char* CryptomatteAOV::get_model() const
+{
+    if (impl->m_layer_type == CryptomatteAOV::CryptomatteType::ObjectNames)
+        return CryptomatteObjectAOVModel;
+    else
+        return CryptomatteMaterialAOVModel;
+}
+
+size_t CryptomatteAOV::get_channel_count() const
+{
+    return 0;
+}
+
+const char** CryptomatteAOV::get_channel_names() const
+{
+    return nullptr;
+}
+
+bool CryptomatteAOV::has_color_data() const
+{
+    return false;
+}
+
+void CryptomatteAOV::create_image(
+    const size_t    canvas_width,
+    const size_t    canvas_height,
+    const size_t    tile_width,
+    const size_t    tile_height,
+    ImageStack&     aov_images)
+{
+    // Create underlying images.
+    const size_t num_channels = (impl->m_num_layers * 2) + 3;
+    impl->m_image.reset(
+        new Image(
+            canvas_width,
+            canvas_height,
+            tile_width,
+            tile_height,
+            num_channels,
+            PixelFormatFloat));
+    impl->m_pixel_samples.resize(canvas_width * canvas_height, WeightMap(impl->m_num_layers));
+    clear_image();
+}
+
+void CryptomatteAOV::clear_image()
+{
+    const size_t num_channels = (impl->m_num_layers * 2) + 3;
```

Comment:
The expression `(impl->m_num_layers * 2) + 3` is used in multiple places, what about factoring it into a method (of `Impl` for instance) and using it instead?

It would also make sense to explain the expression (why the multiplication by 2? Why adding 3).